clippy_lints/src/assertions_on_constants.rs

```diff
@@ -63,7 +65,93 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for AssertionsOnConstants {
             if assert_span.from_expansion() {
                 return;
             }
-            lint_assert_cb(false);
+            if let Some((panic_message, is_true)) = assert_with_message(&cx, e) {
+                if is_true {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        "`assert!(true)` will be optimized out by the compiler",
+                        "remove it",
+                    );
+                } else if panic_message.starts_with("assertion failed: ") {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        "`assert!(false)` should probably be replaced",
+                        "use `panic!()` or `unreachable!()`",
+                    );
+                } else {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        &format!("`assert!(false, \"{}\")` should probably be replaced", panic_message,),
+                        &format!(
+                            "use `panic!(\"{}\")` or `unreachable!(\"{}\")`",
+                            panic_message, panic_message,
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+// fn get_assert_args(snip: String) -> Option<Vec<String>> {
+//
+// }
+
+fn assert_with_message<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) -> Option<(LocalInternedString, bool)> {
```

Comment:
Can you add a doc comment on how the expansion looks like, that you match here, like in
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/format.rs#L171-L181